lib/surface/compiler/ast_translator.ex

```diff
@@ -0,0 +1,281 @@
+defmodule Surface.Compiler.AstTranslator do
+  @behaviour Surface.Compiler.NodeTranslator
+
+  @constructs %{
+    "if" => Surface.Construct.If,
+    "for" => Surface.Construct.For,
+    "template" => Surface.Construct.Template,
+    "slot" => Surface.Construct.Slot
+  }
+
+  @directives [
+    Surface.Directive.ComponentProps,
+    Surface.Directive.Let,
+    Surface.Directive.Hook,
+    Surface.Directive.Values,
+    Surface.Directive.TagAttrs,
+    Surface.Directive.Events,
+    Surface.Directive.Show,
+    Surface.Directive.If,
+    Surface.Directive.For,
+    Surface.Directive.Debug
+  ]
+
+  alias Surface.AST
+  alias Surface.Compiler.Helpers
+  alias Surface.Compiler.Parser
+
+  defguardp is_component(context) when context.type in [AST.Component, AST.MacroComponent]
+  defguardp is_construct(context) when context.type == Surface.Construct
+  defguardp is_subblock(context) when context.type == Surface.Construct.SubBlock
+  defguardp is_tag(context) when context.type in [AST.Tag, AST.VoidTag]
+
+  def handle_comment(state, _comment), do: {state, :ignore}
+
+  def handle_literal(state, value) do
+    {state, %AST.Literal{value: value}}
+  end
+
+  def handle_end(_state, children) do
```

Comment:
my original thought was that it might make sense to collect information while processing the tree and store that in the state, then this would allow a translator to return some additional information alongside the ast. i'm not sure that'll end up being useful, and if not i'll just remove it :-)

Reply:
Interesting idea. I am wondering if we could plug there some static tool analyser like credo that could give to the end-user some useful information about the complexity of the app, best practices, etc.